keras_nlp/src/models/stable_diffusion_3/stable_diffusion_3_backbone.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2024 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import keras
+from keras import layers
+from keras import ops
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.models.backbone import Backbone
+from keras_nlp.src.models.stable_diffusion_3.flow_match_euler_discrete_scheduler import (
+    FlowMatchEulerDiscreteScheduler,
+)
+from keras_nlp.src.models.stable_diffusion_3.mmdit import MMDiT
+from keras_nlp.src.models.stable_diffusion_3.vae_image_decoder import (
+    VAEImageDecoder,
+)
+from keras_nlp.src.utils.keras_utils import standardize_data_format
+
+
+class CLIPProjection(layers.Layer):
+    def __init__(self, hidden_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.hidden_dim = int(hidden_dim)
+
+        self.dense = layers.Dense(
+            hidden_dim,
+            use_bias=False,
+            dtype=self.dtype_policy,
+            name="dense",
+        )
+
+    def build(self, inputs_shape, token_ids_shape):
+        inputs_shape = list(inputs_shape)
+        self.dense.build([None, inputs_shape[-1]])
+
+        # Assign identity matrix to the kernel as default.
+        self.dense._kernel.assign(ops.eye(self.hidden_dim))
+
+    def call(self, inputs, token_ids):
+        indices = ops.expand_dims(
+            ops.cast(ops.argmax(token_ids, axis=-1), "int32"), axis=-1
+        )
+        pooled_output = ops.take_along_axis(inputs, indices[:, :, None], axis=1)
+        pooled_output = ops.squeeze(pooled_output, axis=1)
+        return self.dense(pooled_output)
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "hidden_dim": self.hidden_dim,
+            }
+        )
+        return config
+
+
+@keras_nlp_export("keras_nlp.models.StableDiffusion3Backbone")
+class StableDiffusion3Backbone(Backbone):
```

Comment:
Backbone is looking good! one good thing to test is `model.summary()` and make sure it's looking reasonable, now that we are on a functional model.

Reply:
I have added some helper layers to get a cleaner `model.summary()`.
Now it looks like this:

```bash
Model: "stable_diffusion3_backbone"
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                  ┃ Output Shape              ┃         Param # ┃ Connected to               ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ clip_l_token_ids (InputLayer) │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_g_token_ids (InputLayer) │ (None, None)              │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_l_negative_token_ids     │ (None, None)              │               0 │ -                          │
│ (InputLayer)                  │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_g_negative_token_ids     │ (None, None)              │               0 │ -                          │
│ (InputLayer)                  │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_l (CLIPTextEncoder)      │ [(None, None, 768),       │     123,060,480 │ clip_l_token_ids[0][0],    │
│                               │ (None, None, 768)]        │                 │ clip_l_negative_token_ids… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_g (CLIPTextEncoder)      │ [(None, None, 1280),      │     693,021,440 │ clip_g_token_ids[0][0],    │
│                               │ (None, None, 1280)]       │                 │ clip_g_negative_token_ids… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ num_steps (InputLayer)        │ ()                        │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_1 (Concatenate)   │ (None, None, 2048)        │               0 │ clip_l[0][0], clip_g[0][0] │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_3 (Concatenate)   │ (None, None, 2048)        │               0 │ clip_l[1][0], clip_g[1][0] │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ cast_63 (Cast)                │ (None, None, 768)         │               0 │ clip_l[0][1]               │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ cast_64 (Cast)                │ (None, None, 1280)        │               0 │ clip_g[0][1]               │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ cast_65 (Cast)                │ (None, None, 768)         │               0 │ clip_l[1][1]               │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ cast_66 (Cast)                │ (None, None, 1280)        │               0 │ clip_g[1][1]               │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ scheduler                     │ ()                        │               0 │ num_steps[0][0],           │
│ (FlowMatchEulerDiscreteSched… │                           │                 │ num_steps[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pad (Pad)                     │ (None, None, 4096)        │               0 │ concatenate_1[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pad_2 (Pad)                   │ (None, None, 4096)        │               0 │ concatenate_3[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_l_projection             │ (None, 768)               │         589,824 │ cast_63[0][0],             │
│ (CLIPProjection)              │                           │                 │ clip_l_token_ids[0][0],    │
│                               │                           │                 │ cast_65[0][0],             │
│                               │                           │                 │ clip_l_negative_token_ids… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ clip_g_projection             │ (None, 1280)              │       1,638,400 │ cast_64[0][0],             │
│ (CLIPProjection)              │                           │                 │ clip_g_token_ids[0][0],    │
│                               │                           │                 │ cast_66[0][0],             │
│                               │                           │                 │ clip_g_negative_token_ids… │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ multiply (Multiply)           │ ()                        │               0 │ scheduler[0][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ latents (InputLayer)          │ (None, 100, 100, 16)      │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pad_1 (Pad)                   │ (None, None, 4096)        │               0 │ pad[0][0]                  │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ pad_3 (Pad)                   │ (None, None, 4096)        │               0 │ pad_2[0][0]                │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate (Concatenate)     │ (None, 2048)              │               0 │ clip_l_projection[0][0],   │
│                               │                           │                 │ clip_g_projection[0][0]    │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_2 (Concatenate)   │ (None, 2048)              │               0 │ clip_l_projection[1][0],   │
│                               │                           │                 │ clip_g_projection[1][0]    │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ broadcast_to (BroadcastTo)    │ (None)                    │               0 │ multiply[0][0]             │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_4 (Concatenate)   │ (None, None, 4096)        │               0 │ pad_1[0][0], pad_3[0][0]   │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_6 (Concatenate)   │ (None, 100, 100, 16)      │               0 │ latents[0][0],             │
│                               │                           │                 │ latents[0][0]              │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_5 (Concatenate)   │ (None, 2048)              │               0 │ concatenate[0][0],         │
│                               │                           │                 │ concatenate_2[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ concatenate_7 (Concatenate)   │ (None)                    │               0 │ broadcast_to[0][0],        │
│                               │                           │                 │ broadcast_to[0][0]         │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ diffuser (MMDiT)              │ (None, None, None, 16)    │   2,084,951,104 │ concatenate_4[0][0],       │
│                               │                           │                 │ concatenate_6[0][0],       │
│                               │                           │                 │ concatenate_5[0][0],       │
│                               │                           │                 │ concatenate_7[0][0]        │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ cast_67 (Cast)                │ (None, None, None, 16)    │               0 │ diffuser[0][0]             │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ guidance_scale (InputLayer)   │ ()                        │               0 │ -                          │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ classifier_free_guidance      │ (None, None, None, 16)    │               0 │ cast_67[0][0],             │
│ (ClassifierFreeGuidance)      │                           │                 │ guidance_scale[0][0]       │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ euler_step (EulerStep)        │ (None, 100, 100, 16)      │               0 │ latents[0][0],             │
│                               │                           │                 │ classifier_free_guidance[… │
│                               │                           │                 │ scheduler[0][0],           │
│                               │                           │                 │ scheduler[1][0]            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ latent_calibration            │ (None, 100, 100, 16)      │               0 │ euler_step[0][0]           │
│ (LatentCalibration)           │                           │                 │                            │
├───────────────────────────────┼───────────────────────────┼─────────────────┼────────────────────────────┤
│ decoder (VAEImageDecoder)     │ (None, 800, 800, 3)       │      49,545,475 │ latent_calibration[0][0]   │
└───────────────────────────────┴───────────────────────────┴─────────────────┴────────────────────────────┘
 Total params: 2,952,806,723 (5.50 GB)
 Trainable params: 2,952,806,723 (5.50 GB)
 Non-trainable params: 0 (0.00 B)
```